### ISSUE TITLE
[FR]: Define custom presets friendly name and icon

### ISSUE THREAD

**Issue by s1mpleman** [litinoveweedle/hass-template-climate#27]
### Is your feature request related to a problem? Please describe.

Just a display issue on the custom presets I have that don't match a hass default preset.
![image](https://github.com/user-attachments/assets/97323de0-18ca-4f2e-8ab8-a50452b34d45)

### Describe the solution you'd like

Be able to define the friendly_name, icon for custom presets. Maybe something like the below is possible? 

```
        {
          'off': {
            'name': 'Off',
            'icon': "phu:wiser-thermostat-off",
            'hvac_mode': "off",
            'target_temperature': 5,
          }, 
 ```

### Describe alternatives you've considered

There isn't really any other alternatives I can think of. 

If it's supported by home assistant then that would be awesome to be able to set an icon and friendly name. If not, no trouble!

**Comment by litinoveweedle** [comment]
Ok, the thing is, that I would love to have this functionality as well, but I was not able to crack it. Maybe with your help...

So the icons are not defined in the integration code, but rather in the static [icons.json file](https://developers.home-assistant.io/blog/2024/01/19/icon-translations/). That itself could be handled by the integration creating content this file on its start, but then you would need second HA restart so this file would be taken into account. Not handy not pretty but doable.

The issue was, that when I crated such file as described in the HA documentation it did not worked for me. I tried everything even searched for similar HA issues and found [this one](https://github.com/home-assistant/core/issues/110127). But as you can see I was told, that is should work.

So if you will be able to create workable static icons.json file example I can implement its creation in the integration. You can start with my non working example of the icons.json file from [here](https://github.com/litinoveweedle/hass-template-climate/commit/d7ec2e8e0f26e3072f79ebc60fec196897e13a90)

**Comment by s1mpleman** [comment]
Thank you for the pointers! I have some free time coming up over the Christmas holidays I will have a look into it! I know a little python, more JS and I've been wanting to make some integrations for ages now so will be good to take a look.

**Comment by litinoveweedle** [comment]
Hello good to hear. Just one note. This is not about python as I will gladly do the integration changes myself. 

The icons translation in HA is provided via static JSON file. So I'm theory you can just create mentioned JSON file inside of your integration and reload HA to be able to see new icons. 

Then for sure I will write the functionality which would create that JSON file from the config data as a part of the integration functionality.

But sadly I was unable to get my JSON file working... So if you are going to find a way, how you can create working JSON file manually, I will find a way how to create it from config data, ok?

**Comment by s1mpleman** [comment]
> Hello good to hear. Just one note. This is not about python as I will gladly do the integration changes myself.
> 
> The icons translation in HA is provided via static JSON file. So I'm theory you can just create mentioned JSON file inside of your integration and reload HA to be able to see new icons.
> 
> Then for sure I will write the functionality which would create that JSON file from the config data as a part of the integration functionality.
> 
> But sadly I was unable to get my JSON file working... So if you are going to find a way, how you can create working JSON file manually, I will find a way how to create it from config data, ok?

Thank you for the further clarification! Seemingly easier then! I'm knowledgeable on JSON format so hopefully can figure it out. I'll take a look at the wiser integration as they have custom presets with icons there.

**Comment by s1mpleman** [comment]
I hope you don't mind me collecting data here. 

https://github.com/asantaga/wiserHomeAssistantPlatform/blob/master/custom_components/wiser/icons.json

```
{
    "entity": {
        "climate": {
            "wiser": {
                "state_attributes": {
                    "preset_mode": {
                        "default": "mdi:tune-variant",
                        "state": {
                            "Advance Schedule": "mdi:calendar-arrow-right",
                            "Cancel Overrides": "mdi:close-circle",
                            "Boost 30m": "mdi:fraction-one-half",
                            "Boost 1h": "mdi:numeric-1",
                            "Boost 2h": "mdi:numeric-2",
                            "Boost 3h": "mdi:numeric-3"
                        }
                    }
                }
            }
        }
    },
    "services": {
        "get_schedule": "mdi:calendar-export",
        "set_schedule": "mdi:calendar-import",
        "assign_schedule": "mdi:calendar-check",
        "copy_schedule": "mdi:calendar-blank-multiple",
        "set_device_mode": "mdi:home-switch-outline",
        "boost_heating": "mdi:thermometer-chevron-up",
        "set_opentherm_parameter": "mdi:application-cog-outline"
    }
}
```
```
{
  "entity": {
    "climate": {
      "climate_template": {
        "state_attributes": {
          "preset_mode": {
            "default": "mdi:thermometer-alert",
            "state": {
              "automatic": "mdi:thermometer-auto",
              "comfort": "mdi:thermometer-high",
              "reduced": "mdi:thermometer",
              "protection": "mdi:snowflake-thermometer"
            }
          }
        }
      }
    }
  }
}
```

**Comment by litinoveweedle** [comment]
Not at all. You can also check mine try, which I linked in my first reply. This was already tailored  for the climate template. You can compare both files and check where I made an error, as I was unable to make it work.

**Comment by s1mpleman** [comment]
> Not at all. You can also check mine try, which I linked in my first reply. This was already tailored for the climate template. You can compare both files and check where I made an error, as I was unable to make it work.

Perfect thank you!

**Comment by s1mpleman** [comment]
Nothing wrong with your JSON either way. The only difference with the wiser one is they are referencing using the friendly name of the preset. I gather it isn't something as silly as that? Would be weird if it was. 

I've read through those issues you shared too, but I have no HASS development environment. I will get both of those sorted and get back to you.

**Comment by litinoveweedle** [comment]
You should not need HA development environment "I think". It shall be enough to get HA SSH addon running and create the JSON file inside of the custom_component/template_climate directory and then restart HA. But as I said, I tried that and it was not working.

**Comment by s1mpleman** [comment]
Just going off your comments on the issue I didn't think it would be as simple as doing that but I require the environment for other reasons as well, so may as well get it sorted. 

I did just try creating the json file though and it wouldn't work with either case for the name's as you stated. What's weird is it's leaving all the unrecognized options in lower case too. 

There are some difference's between how your preset modes display in the frontend to the wiser one. Under developer tools for my wiser device I get the following:
`preset_modes: Advance Schedule, Cancel Overrides, Boost 30m, Boost 1h, Boost 2h, Boost 3h`
On my template device that sits over the wiser I get this: 
```
presets: 
  'off':
    hvac_mode: 'off'
    target_temperature: 5
  passive:
    hvac_mode: null
    target_temperature: null
  away:
    hvac_mode: auto
    target_temperature: 14
  holiday:
    hvac_mode: auto
    target_temperature: 7.5
  sleep:
    hvac_mode: auto
    target_temperature: 10
  protection:
    hvac_mode: auto
    target_temperature: 13
  background:
    hvac_mode: heat
    target_temperature: 14.700000000000001
  eco:
    hvac_mode: heat
    target_temperature: 17
  comfort:
    hvac_mode: heat
    target_temperature: 18
  boost:
    hvac_mode: heat
    target_temperature: 19
```

Currently the preset_modes variable isn't filled in from what I can tell from this integration. 

Could it just require that list building?

EDIT:
I'm blind apparently:
`preset_modes: off, passive, away, holiday, sleep, protection, background, eco, comfort, boost`

Not that then...

**Comment by litinoveweedle** [comment]
I do not think that this is the issue as well.

But what it seems to be missing is the _attr_translation_key set to "climate_template" in the integration and also maybe some other attributes. There is note in the developer docs that translation only works for entities [having unique_id](https://developers.home-assistant.io/blog/2022/12/01/entity_translations/) and also some users reported needs to [set _attr_has_entity_name](https://community.home-assistant.io/t/translation-issues/657691/7).

Both of those needs to be set in the Python code. I will try to update my original icons branch with this - you can than use it for further testing.

**Comment by s1mpleman** [comment]
> I do not think that this is the issue as well.
> 
> But what it seems to be missing is the _attr_translation_key set to "climate_template" in the integration and also maybe some other attributes. There is note in the developer docs that translation only works for entities [having unique_id](https://developers.home-assistant.io/blog/2022/12/01/entity_translations/) and also some users reported needs to [set _attr_has_entity_name](https://community.home-assistant.io/t/translation-issues/657691/7).
> 
> Both of those needs to be set in the Python code. I will try to update my original icons branch with this - you can than use it for further testing.

Yeah I'd tried adding the _attr_translation_key="climate_template" to my local installation with no avail. There's something else missing.  I'll see if I can find more time to have a look too. 

Interesting about the unique id and attr_has_entity_name, thanks as I can add the attr_has_entity name but no idea with unqiue id, I thought it already had one. 

I don't think it's anything to do with a strings.json, as wiser integration isn't providing that for the preset attribute. 

This is pretty ridiculous it isn't documented.

**Comment by BirbByte** [comment]
Not exactly the same bug, but same underlying gap, and worth linking here for anyone else who finds this issue.

For the *icon* half specifically: Home Assistant already has a generic mechanism for this — icon translations, specifically the `state_attributes` block in an integration's `icons.json`, keyed by `entity.climate.<translation_key>.state_attributes.<attribute>.state.<value>`. It covers `fan_mode`, `swing_mode`, and `preset_mode` the same way — an integration just needs to (a) set `_attr_translation_key` on the entity, and (b) ship an `icons.json` mapping each state value to an mdi icon. Docs: https://developers.home-assistant.io/docs/core/integration-quality-scale/rules/icon-translations/

I hit this for fan_mode/swing_mode and got it working, but tracking down *why* it wasn't showing took a while: `climate_template`'s `__init__.py` doesn't define a top-level `async_setup()`, so Home Assistant never registers `climate_template` under its bare domain name in `hass.config.components` — only the dotted `climate_template.climate` form. The frontend's icon lookup gates on the bare domain name being present (`isComponentLoaded`, in `ha-attribute-icon`/`getPlatformIcons`), so it silently never even requests `icons.json`, correct or not. A no-op fixes that for good:

```python
async def async_setup(hass, config):
    return True
```

That gets you icons for `preset_mode` the same way as fan/swing — though as a static `icons.json` shipped with the integration, not a per-user YAML field, so it doesn't cover the inline `name`+`icon`+`hvac_mode`+`target_temperature` syntax proposed above. The `friendly_name` part of the original ask seems mostly moot already, since `preset_modes` already accepts arbitrary strings for display.

Fork with the `async_setup` fix + a worked `icons.json` example, if useful:
https://github.com/BirbByte/hass-template-climate

Example of what this looks like:

<img width="573" height="731" alt="Image" src="https://github.com/user-attachments/assets/dc4d3500-080f-4ef7-babc-1fb0bd72885f" />

<img width="494" height="359" alt="Image" src="https://github.com/user-attachments/assets/70299eb3-c61c-4cb6-839b-6d0bd5b6b4e4" />

**Comment by litinoveweedle** [comment]
Hello, this is great. I just tried something (without any guarantees as I am already dead tired and had no time to test it): [icons branch](https://github.com/litinoveweedle/hass-template-climate/tree/icons) shall be able to generate the icons.json file on the fly on the HA restart from the entity configuration. Feature is documented in the README.

**Comment by BirbByte** [comment]
I tried this out on my real setup: three AC entities, each with its own custom icons for fan speed and swing position, using my own names instead of the defaults (like "Quiet" and "Max" instead of what Home Assistant expects by default). After a full restart, the `icons.json` file got created automatically like you said, and the icons show up correctly.

One thing worth knowing: I notice this branch has a fair bit more in it than just the icon feature. Looks like there's also new stuff for setting min/max temperature and humidity through templates. I've only actually tested the icon part, so I can't speak to whether the rest works.

Also, small thing in the README: the note about restarting Home Assistant still mentions something called `icon_translation_key`, but that's not actually a setting anymore in this version, only "icons" is. Probably just needs that leftover word removed.

Would I be correct in assuming these icons changes you've pushed here are likely to make it to the next release once all testing is finalised?

**Comment by litinoveweedle** [comment]
Thank you for catching the docs bug, I will fix it. The translation key is created automatically form entity unique_id or name, so there is no need to configure this.

You are completely correct, the icon branch is built on top of the overhaul branch containing multiple fixes and new features - one fix seems to be required for the HA 2026.09.

Please check the branch if you will have time and feels free to raise PR towards the icons branch. Thank you

**Comment by litinoveweedle** [comment]
So I tested this and I can confirm it works as expected. Thank you once more for finding that piece of puzzle I was unable to solve!

Where I was at it I also implemented similar mechanism for the language translation ;-)

<img width="495" height="615" alt="Image" src="https://github.com/user-attachments/assets/6a67ad4f-cc7c-4cfe-b258-5379654d99de" />

This improvement will be merged into incoming release ASAP. Big thanks once again ;-)
